fix(common): fix immutability for locale_data_api

artem-galas opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Since no new array instance was returned (on `getLocaleMonthNames` / `getLocaleDayNames` methods), after manipulation (e.q `splice`, `push`) on array it's not possible to get full list of localized days/month.

Issue Number: #27003

## What is the new behavior?
Add `Array.from` to  return value from these methods. 
Add general tests for both methods (they didn't have any before), and also add tests for testing bugfix  

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alxhub commented:
I think a better change would be to have this API return `ReadonlyArray<...>`.

artem-galas commented:
Thanks for your review. 
I agree it's better to return `ReadonlyArray`, in that case we freeze these arrays for manipulations.

artem-galas commented:
Test are failing with error
```
Error: Symbol not found for identifier: readonly
```

I don't have any idea what does it mean.
Could you please give me a piece of advise?

ocombe commented:
The issue on here: https://github.com/angular/angular/pull/30397/files#diff-4a042642e9a286cfca1ed035a1fca39aL234
you're using `readonly` instead of `ReadonlyArray`

ocombe commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -231,7 +231,7 @@ export function getLocaleId(locale: string): string {
  */
 export function getLocaleDayPeriods(
-    locale: string, formStyle: FormStyle, width: TranslationWidth): [string, string] {
+    locale: string, formStyle: FormStyle, width: TranslationWidth): readonly[string, string] {
```
should be `ReadonlyArray<string>`

ocombe replied:
actually that makes the type less precise if we use that, not sure what's the correct usage here
@alxhub ?

artem-galas replied:
I agree with.
Maybe this array we can keep without readonly?

ocombe commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -294,7 +293,8 @@ export function getLocaleMonthNames(
  */
-export function getLocaleEraNames(locale: string, width: TranslationWidth): [string, string] {
+export function getLocaleEraNames(
+    locale: string, width: TranslationWidth): readonly[string, string] {
```
should be `ReadonlyArray<string>`

ocombe replied:
actually that makes the type less precise if we use that, not sure what's the correct usage here
@alxhub ?

petebacondarwin commented:
@artem-galas - would you mind rebasing this on master?

petebacondarwin commented:
Can you also tighten up the commit message a bit so that readers of the changelog will understand more what has changed. Perhaps something more like:


```
fix(common): mark locale data arrays as readonly

To discourage developers from mutating the arrays returned
from the following methods, their return types have been marked
as readonly.

* `getLocaleDayPeriods()`
* `getLocaleDayNames()`
* `getLocaleMonthNames()`
* `getLocaleEraNames()`

Fixes #27003

BREAKING CHANGE:
The locale data API has been marked as returning readonly arrays, rather
than mutable arrays, since these arrays are shared across calls to the
API. If you were mutating them (e.g. calling `sort()`, `push()`, `splice()`, etc)
then your code will not longer compile. If you need to mutate the array, you
should now take a copy (e.g. by calling `slice()`) and mutate the copy.
```

Commit:
fix(common): mark locale data arrays as readonly

To discourage developers from mutating the arrays returned
from the following methods, their return types have been marked
as readonly.

* `getLocaleDayPeriods()`
* `getLocaleDayNames()`
* `getLocaleMonthNames()`
* `getLocaleEraNames()`

Fixes #27003

BREAKING CHANGE:
The locale data API has been marked as returning readonly arrays, rather
than mutable arrays, since these arrays are shared across calls to the
API. If you were mutating them (e.g. calling `sort()`, `push()`, `splice()`, etc)
then your code will not longer compile. If you need to mutate the array, you
should now take a copy (e.g. by calling `slice()`) and mutate the copy.

petebacondarwin commented:
I just realised that this will probably fail due to the public API change. We will need to update the public API "golden files"...
The CI test failures will say what commands need to be run to do this.

petebacondarwin reviewed (approved):
As this is a breaking change to the public API, it could only land in 11.0.0

petebacondarwin commented:
I think it is worth running a presubmit now to see how many Google projects would be broken by this change.

petebacondarwin commented on goldens/public-api/common/common.d.ts:
```diff
@@ -61,13 +61,13 @@ export declare function getLocaleDateFormat(locale: string, width: FormatWidth):
+export declare function getLocaleDayNames(locale: string, formStyle: FormStyle, width: TranslationWidth): ReadonlyArray<string>;
 
-export declare function getLocaleDayPeriods(locale: string, formStyle: FormStyle, width: TranslationWidth): [string, string];
+export declare function getLocaleDayPeriods(locale: string, formStyle: FormStyle, width: TranslationWidth): ReadonlyArray<string>;
```
Actually I just realised that we can do this:

```
Readonly<[string, string]>
```

artem-galas replied:
Oh, yes. Good catch!

artem-galas commented:
Thanks for your review!

> As this is a breaking change to the public API, it could only land in 11.0.0

I'll try to keep it up to date as long as possible.

mhevery commented:
[presubmit](http://test/OCL:320982824:BASE:320982860:1594661274400:94d1958a)

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

petebacondarwin reviewed (approved):
Reviewed-for: public-api
Reviewed-for: fw-i18n

jelbourn commented:
@mhevery could you review for fw-core?

AndrewKushnir reviewed (approved):
LGTM 👍 

Reviewed-for: fw-core

AndrewKushnir commented:
**Note to Caretakers**: since this is a breaking change, we'd need to run Global presubmit as well once master branch is available for v11 changes.

AndrewKushnir reviewed (approved):
Adding LGTM for the `public-api` group.

Reviewed-for: public-api

atscott reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
Hi @artem-galas, the 10.1.0 version was released last week and master branch is now available for changes that target next major release (v11). Could you please rebase this PR, so that we re-run tests with the current version of master branch? Please let us know (in this thread) once this PR is rebased, so that we re-run tests in Google's codebase. Thank you.

mhevery reviewed (approved):
Reviewed-for: fw-core

artem-galas commented:
Branch is rebased to latest master.

AndrewKushnir commented:
[Presubmit](http://test/OCL:330831366:BASE:330831388:1599696924289:25883940).

AndrewKushnir commented:
FYI, presubmits (including a global one) went well, this PR is ready for merge.
